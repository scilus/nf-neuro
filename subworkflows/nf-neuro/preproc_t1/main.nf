// ** Importing modules from nf-neuro ** //
include { DENOISING_NLMEANS } from '../../../modules/nf-neuro/denoising/nlmeans/main'
include { PREPROC_N4 } from '../../../modules/nf-neuro/preproc/n4/main'
include { IMAGE_RESAMPLE } from '../../../modules/nf-neuro/image/resample/main'
include { BETCROP_ANTSBET } from '../../../modules/nf-neuro/betcrop/antsbet/main'
include { BETCROP_SYNTHBET} from '../../../modules/nf-neuro/betcrop/synthbet/main'
include { IMAGE_CROPVOLUME as IMAGE_CROPVOLUME_T1 } from '../../../modules/nf-neuro/image/cropvolume/main'
include { IMAGE_CROPVOLUME as IMAGE_CROPVOLUME_MASK } from '../../../modules/nf-neuro/image/cropvolume/main'

params.run_synthbet = false

workflow PREPROC_T1 {

    take:
        ch_image            // channel: [ val(meta), image ]
        ch_template         // channel: [ val(meta), template ]                              , optional
        ch_probability_map  // channel: [ val(meta), probability-map, mask, initial-affine ] , optional
        ch_mask_nlmeans     // channel: [ val(meta), mask ]                                  , optional
        ch_ref_n4           // channel: [ val(meta), ref, ref-mask ]                         , optional
        ch_ref_resample     // channel: [ val(meta), ref ]                                   , optional
        ch_weights          // channel: [ val(meta), weights ]                               , optional

    main:

        ch_versions = Channel.empty()

        if ( params.run_denoising ) {

            // ** Denoising ** //
            // Result : [ meta, image, mask | [] ]
            //  Steps :
            //   - join [ meta, image, mask | null ]
            //   - map  [ meta, image, mask | [] ]
            ch_nlmeans = ch_image
                .join(ch_mask_nlmeans, remainder: true)
                .map{ it[0..1] + [it[2] ?: []] }

            DENOISING_NLMEANS ( ch_nlmeans )
            ch_versions = ch_versions.mix(DENOISING_NLMEANS.out.versions.first())
            image_nlmeans = DENOISING_NLMEANS.out.image
        }
        else {
            image_nlmeans = ch_image
        }

        if ( params.run_N4 ) {
            // ** N4 correction ** //
            // Result : [ meta, image, reference | [], mask | [] ]
            //  Steps :
            //   - join [ meta, image ] + [ reference, mask ] | [ reference, null ] | [ null ]
            //   - map  [ meta, image, reference | [], mask | [] ]
            //   - join [ meta, image, reference | [], mask | [], nlmeans-mask | null ]
            //   - map  [ meta, image, reference | [], mask | [] ]
            ch_N4 = image_nlmeans
                .join(ch_ref_n4, remainder: true)
                .map{ it[0..1] + [it[2] ?: [], it[3] ?: []] }
                .join(ch_mask_nlmeans, remainder: true)
                .map{ it[0..2] + [it[3] ?: it[4] ?: []] }

            PREPROC_N4 ( ch_N4 )
            ch_versions = ch_versions.mix(PREPROC_N4.out.versions.first())
            image_N4 = PREPROC_N4.out.image
        }
        else {
            image_N4 = image_nlmeans
        }

        if ( params.run_resampling ) {
            // ** Resampling ** //
            // Result : [ meta, image, reference | [] ]
            //  Steps :
            //   - join [ meta, image, reference | null ]
            //   - map  [ meta, image, reference | [] ]
            ch_resampling = image_N4
                .join(ch_ref_resample, remainder: true)
                .map{ it[0..1] + [it[2] ?: []] }

            IMAGE_RESAMPLE ( ch_resampling )
            ch_versions = ch_versions.mix(IMAGE_RESAMPLE.out.versions.first())
            image_resample = IMAGE_RESAMPLE.out.image
        }
        else {
            image_resample = image_N4
        }

        if ( params.run_synthbet ) {
            // ** SYNTHBET ** //
            // Result : [ meta, image, weights | [] ]
            //  Steps :
            //   - join [ meta, image, weights | null ]
            //   - map  [ meta, image, weights | [] ]
            ch_bet = image_resample
                .join(ch_weights, remainder: true)
                .map{ it[0..1] + [it[2] ?: []] }

            BETCROP_SYNTHBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_SYNTHBET.out.versions.first())

            // ** Setting BET output ** //
            image_bet = BETCROP_SYNTHBET.out.bet_image
            mask_bet = BETCROP_SYNTHBET.out.brain_mask
        }
        else if ( params.run_ants_bet ) {
            // ** ANTSBET ** //
            // The template and probability maps are mandatory if running antsBET. Since the
            // error message from nextflow when they are absent is either non-informative or
            // missing, we use ifEmpty to provide a more informative one.
            ch_bet = image_resample
                .join(ch_template.ifEmpty{ error("ANTS BET needs a template") })
                .join(ch_probability_map.ifEmpty{ error("ANTS BET needs a tissue probability map") })

            BETCROP_ANTSBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_ANTSBET.out.versions.first())

            // ** Setting BET output ** //
            image_bet = BETCROP_ANTSBET.out.t1
            mask_bet = BETCROP_ANTSBET.out.mask
        }
        else{
            image_bet = image_resample
            mask_bet = Channel.empty()
        }

        if ( params.run_crop ) {
            // ** Crop image ** //
            ch_crop = image_bet
                .map{ it + [[]] }

            IMAGE_CROPVOLUME_T1 ( ch_crop )
            ch_versions = ch_versions.mix(IMAGE_CROPVOLUME_T1.out.versions.first())
            image_crop = IMAGE_CROPVOLUME_T1.out.image
            bbox = IMAGE_CROPVOLUME_T1.out.bounding_box

            // ** Crop mask ** //
            ch_crop_mask = mask_bet
                .join(IMAGE_CROPVOLUME_T1.out.bounding_box)

            IMAGE_CROPVOLUME_MASK ( ch_crop_mask )
            ch_versions = ch_versions.mix(IMAGE_CROPVOLUME_MASK.out.versions.first())
            mask_crop = IMAGE_CROPVOLUME_MASK.out.image
        }
        else{
            image_crop = image_bet
            mask_crop = Channel.empty()
            bbox = Channel.empty()
        }

    emit:
        t1_final        = image_crop                    // channel: [ val(meta), t1-preprocessed ]
        mask_final      = mask_crop                     // channel: [ val(meta), t1-mask ]
        image_nlmeans   = image_nlmeans                 // channel: [ val(meta), t1-after-denoise ]
        image_N4        = image_N4                      // channel: [ val(meta), t1-after-unbias ]
        image_resample  = image_resample                // channel: [ val(meta), t1-after-resample ]
        image_bet       = image_bet                     // channel: [ val(meta), t1-after-bet ]
        mask_bet        = mask_bet                      // channel: [ val(meta), intermediary-mask ]
        crop_box        = bbox                          // channel: [ val(meta), bounding-box ]
        versions        = ch_versions                   // channel: [ versions.yml ]
}
