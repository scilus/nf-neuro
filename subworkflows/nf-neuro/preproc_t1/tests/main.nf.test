nextflow_workflow {

    name "Test Subworkflow PREPROC_T1"
    script "../main.nf"
    workflow "PREPROC_T1"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/preproc_t1"

    tag "denoising/nlmeans"
    tag "preproc/n4"
    tag "image/resample"
    tag "betcrop/antsbet"
    tag "betcrop/synthbet"
    tag "image/cropvolume"

    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "antsbet.zip" , "T1w.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("preproc_t1_antsbet_error") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[4] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[6] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                """
            }
        }

        then {
            assert workflow.failed
        }
    }

    test("preproc_t1_antsbet") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz"),
                        [],
                        []
                    ]}
                input[3] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[4] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[6] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    niftiMD5SUM(workflow.out.image_nlmeans.get(0).get(1)),
                    niftiMD5SUM(workflow.out.image_N4.get(0).get(1)),
                    niftiMD5SUM(workflow.out.image_resample.get(0).get(1)),
                    file(workflow.out.image_bet.get(0).get(1)).name,
                    file(workflow.out.mask_bet.get(0).get(1)).name,
                    file(workflow.out.crop_box.get(0).get(1)).name,
                    file(workflow.out.mask_final.get(0).get(1)).name,
                    file(workflow.out.t1_final.get(0).get(1)).name,
                    workflow.out.versions
                ).match()}
            )
        }
    }

    test("preproc_t1_synthbet") {
        config "./nextflow_synthbet.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[3] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        [],
                        []
                    ]}
                input[4] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[6] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    niftiMD5SUM(workflow.out.image_nlmeans.get(0).get(1)),
                    niftiMD5SUM(workflow.out.image_N4.get(0).get(1)),
                    niftiMD5SUM(workflow.out.image_resample.get(0).get(1)),
                    niftiMD5SUM(workflow.out.image_bet.get(0).get(1)),
                    niftiMD5SUM(workflow.out.mask_bet.get(0).get(1)),
                    workflow.out.crop_box,
                    niftiMD5SUM(workflow.out.mask_final.get(0).get(1)),
                    niftiMD5SUM(workflow.out.t1_final.get(0).get(1)),
                    workflow.out.versions
                ).match()}
            )
        }
    }

    test("preproc_t1_quick") {
        config "./nextflow_quick.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[3] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        [],
                        []
                    ]}
                input[4] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[6] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("preproc_t1_skip_all") {
        config "./nextflow_skip_all.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[3] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        [],
                        []
                    ]}
                input[4] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        [],
                        []
                    ]}
                input[5] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                input[6] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        []
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
