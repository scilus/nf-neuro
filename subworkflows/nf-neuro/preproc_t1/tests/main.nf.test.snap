{
    "preproc_t1_quick": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "1": [

                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__denoised.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__denoised.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_resampled.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_resampled.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "6": [

                ],
                "7": [

                ],
                "8": [
                    "versions.yml:md5,bdd934b4b8456060c36d6d97e4f30740",
                    "versions.yml:md5,bf4dd58c38dd4863ebfb9e78a94c3a20",
                    "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
                ],
                "crop_box": [

                ],
                "image_N4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__denoised.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_resampled.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__denoised.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_resampled.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "mask_bet": [

                ],
                "mask_final": [

                ],
                "t1_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,c507c9182cc410c298fad4a03540c0c9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,bdd934b4b8456060c36d6d97e4f30740",
                    "versions.yml:md5,bf4dd58c38dd4863ebfb9e78a94c3a20",
                    "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-14T10:02:29.972139887"
    },
    "preproc_t1_synthbet": {
        "content": [
            "test__denoised.nii.gz:md5:header,a7ee0e819007aba98d14f7a145b550e6,data,2e21433e2bcd4de2a6b6167c6233cd40",
            "test__image_n4.nii.gz:md5:header,e7cfbd06624321d70cbd667a77315ba3,data,a81e98f32ed963c098ccb07486101898",
            "test_resampled.nii.gz:md5:header,7628a07204938d640c3530fa3d76d2b7,data,a81e98f32ed963c098ccb07486101898",
            "test__bet_image.nii.gz:md5:header,38a09a5addf7c1f13dae1121c562f3b5,data,74f76699e09839809038875af1d9ae6c",
            "test__brain_mask.nii.gz:md5:header,8fc2b4ae979d881623dfd34d377d437d,data,2d61506dc4ab2f8093b731474ae7e45c",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_t1_cropped_bbox.pkl:md5,522d2c44d3ad1058ea77457a263e39c8"
                ]
            ],
            "test_cropped.nii.gz:md5:header,efab188f3700b5b29d4b4ef99cec1295,data,e551be653d402018fd73a4f708b1641e",
            "test_t1_cropped.nii.gz:md5:header,efab188f3700b5b29d4b4ef99cec1295,data,8a93b9d76ead7dae4af4792b9cf70479",
            [
                "versions.yml:md5,318cabe934be45528a25f52083d9c90d",
                "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                "versions.yml:md5,bdd934b4b8456060c36d6d97e4f30740",
                "versions.yml:md5,be3dbb0ac2589ad263d583018f339102",
                "versions.yml:md5,bf4dd58c38dd4863ebfb9e78a94c3a20",
                "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-25T18:23:03.990173398"
    },
    "preproc_t1_skip_all": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "1": [

                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "6": [

                ],
                "7": [

                ],
                "8": [

                ],
                "crop_box": [

                ],
                "image_N4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "mask_bet": [

                ],
                "mask_final": [

                ],
                "t1_final": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "T1w.nii.gz:md5,ce10054d30c0a0753c619d67d811fe32"
                    ]
                ],
                "versions": [

                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-14T21:44:26.607236449"
    },
    "preproc_t1_antsbet": {
        "content": [
            "test__denoised.nii.gz:md5:header,a7ee0e819007aba98d14f7a145b550e6,data,2e21433e2bcd4de2a6b6167c6233cd40",
            "test__image_n4.nii.gz:md5:header,e7cfbd06624321d70cbd667a77315ba3,data,a81e98f32ed963c098ccb07486101898",
            "test_resampled.nii.gz:md5:header,7628a07204938d640c3530fa3d76d2b7,data,a81e98f32ed963c098ccb07486101898",
            "test__t1_bet.nii.gz",
            "test__t1_bet_mask.nii.gz",
            "test_t1_cropped_bbox.pkl",
            "test_cropped.nii.gz",
            "test_t1_cropped.nii.gz",
            [
                "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                "versions.yml:md5,bdd934b4b8456060c36d6d97e4f30740",
                "versions.yml:md5,be3dbb0ac2589ad263d583018f339102",
                "versions.yml:md5,bf4dd58c38dd4863ebfb9e78a94c3a20",
                "versions.yml:md5,da278daafbe3afa8454021e2716dd205",
                "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-12-12T15:50:29.838128539"
    }
}
